sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -62,6 +63,25 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
   }
 
   /**
+   * Infers an additional set of constraints from a given set of equality constraints.
+   * For e.g., if an operator has constraints of the form (`a = 5`, `a = b`), this returns an
+   * additional constraint of the form `b = 5`
+   */
+  private def inferAdditionalConstraints(constraints: Set[Expression]): Set[Expression] = {
+    constraints.map {
+      case eq @ EqualTo(l: Attribute, r: Attribute) =>
+        (constraints -- Set(eq)).map(_ transform {
+          case a: Attribute if a.semanticEquals(l) => r
+        }).union(
+          (constraints -- Set(eq)).map(_ transform {
+            case a: Attribute if a.semanticEquals(r) => l
+          }))
+      case _ =>
+        Set.empty[Expression]
+    }.foldLeft(Set.empty[Expression])(_ union _) -- constraints
```

Comment:
maybe not foldLeft -- most of the time fold is used it can be rewritten into something imperative that is simpler (and a lot faster)